### PR TITLE
Rewrite Preprocessor

### PR DESCRIPTION
**Pull request by donglihe-hub**
### Highlights
1. Separate loading from preprocessing (now closer to scikit-learn style)
2. Preprocessor now accepts user-defined tf-idf params
3. read_*_format() becomes private. Users have only one way to load data, i.e., load_dataset().
4. Thanks to 3., text and labels can only be access by "x" and "y", respectively, no more "text" or "labels". 
5. Update tutorials

**To-do:**
1. Further investigate if the preprocessor can be embedded into sklearn's pipeline

### Demo for Loading Data and Preprocessing:
**Before:**
from libmultilabel import Preprocessor
\# Load & Preprocessing
preprocessor = Preprocessor(data_format="txt", include_test_labels=False, remove_no_label_dat=False)
datasets = preprocessor.load_data("data/eur-lex/train.txt", "data/eur-lex/test.txt")

**After:**
from libmultilabel.linear import load_dataset, Preprocessor
\# Load
datasets = load_dataset(data_format="txt", train_path="data/eur-lex/train.txt", test_path="data/eur-lex/test.txt")
\# Preprocessing
preprocessor = Preprocessor(include_test_labels=False, remove_no_label_dat=False, tfidf_params = {})
preprocessor.fit_transform(datasets)

## Test CLI & API (`bash tests/autotest.sh`)
Test APIs used by main.py.
- [ ] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
If any new APIs are added, please check if the description of the APIs is added to API document. 
- [x] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_changed_document.sh`)
If any APIs in quickstarts or tutorials are modified, please run this test to check if the current examples can run correctly after the modified APIs are released.

FAILED  python3 docs/examples/plot_linear_gridsearch_tutorial.py
FAILED  python3 docs/examples/plot_dataset_tutorial.py
PASSED  python3 docs/examples/plot_linear_quickstart.py
PASSED  python3 docs/examples/plot_KimCNN_quickstart.py
PASSED  python3 docs/examples/plot_bert_quickstart.py

### CODE REVIEW

File: libmultilabel/linear/preprocessor.py [hunk 1]
```diff
@@ -1,216 +1,124 @@
...
+        self.data_format = None
+        self.vectorizer = None
+        self.binarizer = None
+
+    def fit(self, dataset):
+        """Fit the raw data.
+
+        Parameters
+        ----------
+        dataset : The training and test dataset along with labels with keys 'train', 'test', and labels respectively.
```
Comment (cjlin1) [inline libmultilabel/linear/preprocessor.py:39]:
```suggestion
        dataset : The training and test datasets along with labels with keys 'train', 'test', and labels respectively.
```

File: docs/examples/plot_linear_quickstart.py [hunk 2]
```diff
@@ -13,9 +13,9 @@
 ######################################################################
 # To start, we need to read and preprocess the input data:
 
-preprocessor = linear.Preprocessor(data_format="txt")
-
-datasets = preprocessor.load_data("data/rcv1/train.txt", "data/rcv1/test.txt")
+datasets = linear.load_dataset("txt", "data/rcv1/train.txt", "data/rcv1/test.txt")
+preprocessor = linear.Preprocessor()
+datasets = preprocessor.fit_transform(datasets)
```
Comment (cjlin1) [inline docs/examples/plot_linear_quickstart.py:18]: users may not be familiar with the "fit" and "transform" names used by sklearn. May need more explanation

File: libmultilabel/linear/preprocessor.py [hunk 3]
```diff
@@ -1,216 +1,124 @@
...
-            raise ValueError(f"invalid svm format at line {i+1} of the file '{file_path}'")
-
-    prob_x = scipy.frombuffer(prob_x, dtype="d")
-    col_idx = scipy.frombuffer(col_idx, dtype="l")
-    row_ptr = scipy.frombuffer(row_ptr, dtype="l")
-    prob_x = sparse.csr_matrix((prob_x, col_idx, row_ptr))
-
-    return (prob_y, prob_x)
+    def fit_transform(self, dataset):
+        """Fit the raw data and transform it.
```
Comment (cjlin1) [inline libmultilabel/linear/preprocessor.py:111]: more explanation is needed

File: libmultilabel/linear/preprocessor.py [hunk 4]
```diff
@@ -1,216 +1,124 @@
...
-        if self.data_format in {"txt", "dataframe"}:
-            data = self._load_text(training_data, test_data, eval)
-        elif self.data_format == "svm":
-            data = self._load_svm(training_data, test_data, eval)
-
-        if "train" in data:
-            num_labels = data["train"]["y"].getnnz(axis=1)
+        if self.binarizer is None:
+            raise AttributeError("Preprecessor has not been fitted.")
+        ds = copy.deepcopy(dataset)
```
Comment (cjlin1) [inline libmultilabel/linear/preprocessor.py:74]: why we cannot use dataset and need this deepcopy?

Comment (donglihe-hub) [reply libmultilabel/linear/preprocessor.py:74]: Preprocessor follows the style of scikit-learn. Thus, I think it's the dict input should be deep-copied to avoid any modification in the original object.

File: docs/examples/plot_linear_quickstart.py [hunk 5]
```diff
@@ -50,9 +50,9 @@
 # ``preds`` holds the decision values, i.e. the raw values
 # outputted by the model. To transform it into predictions,
 # you can apply the API ``get_positive_labels`` to get predicted labels and their corresponding scores
-# by using ``label_mapping`` in ``preprocessor`` and ``preds`` from the last step.
+# by using ``classes`` in ``preprocessor.binarizer`` and ``preds`` from the last step.
```
[the remark this answers is not in the document]
Comment (donglihe-hub) [reply docs/examples/plot_linear_quickstart.py:53]: We can of course change back to the original usage. Actually what I'm thinking about here is hiding the attribute and writing a method to return it.

File: libmultilabel/linear/preprocessor.py (hunk 4 above)
Comment (Sinacam) [reply libmultilabel/linear/preprocessor.py:74]: Either way, `ds` is not a good name.
```suggestion
        dataset = copy.deepcopy(dataset)
```
or
```suggestion
        dataset_copy = copy.deepcopy(dataset)
```

File: libmultilabel/linear/preprocessor.py [hunk 6]
```diff
@@ -1,216 +1,124 @@
...
+            and self.vectorizer is not None
+        ):
+            if ds.get("train") is not None:
+                ds["train"]["x"] = self.vectorizer.transform(ds["train"]["x"])
+            if ds.get("test") is not None:
+                ds["test"]["x"] = self.vectorizer.transform(ds["test"]["x"])
+        # transform a collection of raw labels to a binary matrix
+        if ds.get("train") is not None:
+            ds["train"]["y"] = self.binarizer.transform(ds["train"]["y"]).astype("d")
+        if ds.get("test") is not None:
```
Comment (Sinacam) [inline libmultilabel/linear/preprocessor.py:87]:
```suggestion
        if "test" in ds:
```

File: libmultilabel/linear/preprocessor.py [hunk 7]
```diff
@@ -1,216 +1,124 @@
...
-            data = self._load_svm(training_data, test_data, eval)
-
-        if "train" in data:
-            num_labels = data["train"]["y"].getnnz(axis=1)
+        if self.binarizer is None:
+            raise AttributeError("Preprecessor has not been fitted.")
+        ds = copy.deepcopy(dataset)
+        # transform a collection of raw text to a matrix of TF-IDF features
+        if (
+            all(f in {"txt", "dataframe"} for f in (self.data_format, ds["data_format"]))
```
Comment (Sinacam) [inline libmultilabel/linear/preprocessor.py:77]: User error checks moved earlier.
```suggestion
            if ds["data_format"] in {"txt", "dataframe"}:
```

File: libmultilabel/linear/preprocessor.py [hunk 8]
```diff
@@ -1,216 +1,124 @@
...
+        self.binarizer = None
+
+    def fit(self, dataset):
+        """Fit the raw data.
+
+        Parameters
+        ----------
+        dataset : The training and test dataset along with labels with keys 'train', 'test', and labels respectively.
+            The dataset has keys 'x' for input features and 'y' for labels. It also contains 'data_format' to indicate
+            the data format used.
```
Comment (Sinacam) [inline libmultilabel/linear/preprocessor.py:41]: The wording does not make it clear these are preconditions.
```suggestion
            The dataset must have keys 'x' for input features, 'y' for labels and 'data_format' to indicate
            the data format used.
```

File: libmultilabel/linear/preprocessor.py [hunk 9]
```diff
@@ -1,216 +1,124 @@
...
-    prob_x = sparse.csr_matrix((prob_x, col_idx, row_ptr))
-
-    return (prob_y, prob_x)
+    def fit_transform(self, dataset):
+        """Fit the raw data and transform it.
+
+        Args:
+            dataset : The training and test dataset along with labels with keys 'train', 'test', and labels respectively.
+            The dataset has keys 'x' for input features and 'y' for labels. It also contains 'data_format' to indicate
+            the data format used.
```
Comment (Sinacam) [inline libmultilabel/linear/preprocessor.py:116]: Same as earlier.
```suggestion
            The dataset must have keys 'x' for input features, 'y' for labels and 'data_format' to indicate
            the data format used.
```

File: libmultilabel/linear/preprocessor.py [hunk 10]
```diff
@@ -1,216 +1,124 @@
...
+
+        Returns:
+            dict[str, dict[str, sparse.csr_matrix]]: The transformed dataset.
+        """
+        return self.fit(dataset).transform(dataset)
+
+    def inverse_transform_label(self, y):
```
Comment (Sinacam) [inline libmultilabel/linear/preprocessor.py:123]: Needs explanation.

File: docs/examples/plot_dataset_tutorial.py [hunk 11]
```diff
@@ -24,10 +24,10 @@
 # We choose a multi-label set ``emoji`` from ``tweet_eval`` in this example.
 # The data set can be loaded by the following code.
 
-data_sets = dict()
-data_sets["train"] = load_dataset("tweet_eval", "emoji", split="train")
-data_sets["val"] = load_dataset("tweet_eval", "emoji", split="validation")
-data_sets["test"] = load_dataset("tweet_eval", "emoji", split="test")
+hf_datasets = dict()
```
[the remark this answers is not in the document]
Comment (Eleven1Liu) [reply docs/examples/plot_dataset_tutorial.py:27]: huggingface
